swarm/storage/resource.go

```diff
@@ -465,78 +438,80 @@ func parseUpdate(blob []byte) (period uint32, version uint32, ensname []byte, da
-		return nil, fmt.Errorf("Invalid resource")
+	// an update can be only one chunk long
+	datalimit := self.chunkSize() - int64(sigoffset-len(name)-8)
+	if int64(len(data)) > datalimit {
+		return nil, fmt.Errorf("Data overflow: %d / %d bytes", len(data), datalimit)
 	}
 
 	// get our blockheight at this time and the next block of the update period
 	currentblock, err := self.getBlock()
 	if err != nil {
 		return nil, err
 	}
-	nextperiod := getNextPeriod(resource.startBlock, currentblock, resource.frequency)
+	nextperiod := getNextPeriod(rsrc.startBlock, currentblock, rsrc.frequency)
 
 	// if we already have an update for this block then increment version
+	// (resource object MUST be in sync for version to be correct)
 	var version uint32
 	if self.hasUpdate(name, nextperiod) {
-		version = resource.version
+		version = rsrc.version
 	}
 	version++
 
-	// prepend version and period to allow reverse lookups
-	// data header length does NOT include the header length prefix bytes themselves
-	headerlength := uint16(len(resource.nameHash) + 4 + 4)
-	fulldata := make([]byte, int(headerlength)+2+len(data))
+	// calculate the chunk key
+	key := self.resourceHash(nextperiod, version, rsrc.nameHash)
 
-	cursor := 0
-	binary.LittleEndian.PutUint16(fulldata, headerlength)
-	cursor += 2
-
-	binary.LittleEndian.PutUint32(fulldata[cursor:], nextperiod)
-	cursor += 4
+	var signature *Signature
+	if self.validator != nil {
+		// sign the data hash with the key
+		digest := self.keyDataHash(key, data)
+		sig, err := self.validator.sign(digest)
+		if err != nil {
+			return nil, err
+		}
+		signature = &sig
 
-	binary.LittleEndian.PutUint32(fulldata[cursor:], version)
-	cursor += 4
+		// get the address of the signer (which also checks that it's a valid signature)
+		addr, err := getAddressFromDataSig(digest, *signature)
+		if err != nil {
+			return nil, fmt.Errorf("Invalid data/signature: %v", err)
+		}
 
-	copy(fulldata[cursor:], resource.nameHash[:])
-	cursor += len(resource.nameHash)
+		// check if the signer has access to update
+		ok, err := self.validator.checkAccess(name, addr)
+		if err != nil {
+			return nil, err
+		} else if !ok {
+			return nil, fmt.Errorf("Address %x does not have access to update %s", addr, name)
+		}
+	}
 
-	copy(fulldata[cursor:], data)
+	chunk := newUpdateChunk(key, signature, nextperiod, version, name, data)
 
-	// create the update chunk and send it
-	key := self.resourceHash(resource.nameHash, nextperiod, version)
-	chunk := NewChunk(key, nil)
-	chunk.SData, err = self.signContent(fulldata)
-	if err != nil {
-		return nil, err
-	}
-	chunk.Size = int64(len(fulldata))
+	// send the chunk
 	self.Put(chunk)
-	log.Trace("resource update", "name", resource.name, "key", key, "currentblock", currentblock, "lastperiod", nextperiod, "version", version, "data", chunk.SData)
+	log.Trace("resource update", "name", name, "key", key, "currentblock", currentblock, "lastperiod", nextperiod, "version", version, "data", chunk.SData)
```

Comment:
Not sure how to, as the chunk waitgroup member is private?

Anyway, the `resource.updated` isn't set, so any subsequent `Get` will block until it can be retrieved...

Reply:
`chunk.WaitToStore(ctx)` it will be, now it is


```
select {
  case <-chunk.dbStored:
  case <- timeout.C:
}
```

Reply:
Is this really good form; to refer to an unexported member of a struct?